src/util/result.h

```diff
@@ -8,16 +8,146 @@
 #include <attributes.h>
 #include <util/translation.h>
 
+#include <memory>
+#include <optional>
+#include <tuple>
+#include <utility>
 #include <variant>
+#include <vector>
 
 namespace util {
+namespace detail {
+//! Empty string list
+const std::vector<bilingual_str> EMPTY_LIST{};
 
+//! Helper function to join messages in space separated string.
+bilingual_str JoinMessages(const std::vector<bilingual_str>& errors, const std::vector<bilingual_str>& warnings);
+
+//! Helper function to move messages from one vector to another.
+void MoveMessages(std::vector<bilingual_str>& src, std::vector<bilingual_str>& dest);
+
+//! Error information only allocated if there are errors or warnings.
+template <typename F>
+struct ErrorInfo {
+    std::optional<std::conditional_t<std::is_same_v<F, void>, std::monostate, F>> failure{};
+    std::vector<bilingual_str> errors;
+    std::vector<bilingual_str> warnings;
+};
+
+//! Result base class which is inherited by Result<T, F>.
+//! T is the type of the success return value, or void if there is none.
+//! F is the type of the failure return value, or void if there is none.
+template <typename T, typename F>
+class ResultBase;
```

Comment:
I think it's a little confusing.
If I understand correctly, `template <typename F> class ResultBase<void, F>` needs to know about `template <typename T, typename F> class ResultBase;` but the latter can only be declared after the former because it is a derived class. Therefore, a forward declaration is required.

If the `ResultBase` classes shouldn't be used anywhere outside of `result.h` and their only purpose is to be the base class of `Result`, I would suggest merging them all into a single class.

Reply:
Hmm, I wonder if you can say more about what is confusing or misleading. This is just a forward declaration for a template class.

 It is true that the reason for the forward declaring `ResultBase<T, F>` is to allow it to inherit from `ResultBase<void, F>`. So `ResultBase<T, F>` must be defined after `ResultBase<void, F>`, but declared before it.

But this is a pretty standard thing for C and C++ code. Definition of one thing 1 depends on declaration of thing 2, so thing 2 needs to be forward declared. It can happen for normal classes and functions as well as templates.

> If the ResultBase classes shouldn't be used anywhere outside of result.h and their only purpose is to be the base class of Result, I would suggest merging them all into a single class.

This isn't easily possible because `Result<void, F>` inherits directly from `detail::ResultBase<void, F>` and does **not** inherit from `detail::ResultBase<T, F>`. It does not have an `m_value` member or `value()` functions or a dereferencing operator. If the result type `T` is void the `Result` class doesn't hold a value and can't be dereferenced.